File: app-presentation/src/main/java/com/zxjdev/smile/presentation/infrastucture/splash/SplashPresenter.java
package com.zxjdev.smile.presentation.infrastucture.splash;

import android.os.Handler;

import com.zxjdev.smile.domain.authorization.usecase.AutoLogin;
import com.zxjdev.smile.presentation.common.DefaultSubscriber;
import com.zxjdev.smile.presentation.common.util.ui.ErrorMessagePrinter;

import javax.inject.Inject;

public class SplashPresenter implements SplashContract.Presenter {

  @Inject AutoLogin autoLogin;
  private SplashContract.View view;
  @Inject ErrorMessagePrinter errorMessagePrinter;
  private Handler handler;

  @Inject
  SplashPresenter() {
    this.handler = new Handler();
  }

  @Override
  public void handleAutoLogin() {
    autoLogin.execute(new DefaultSubscriber<Boolean>(errorMessagePrinter) {
      @Override
      public void onNext(Boolean data) {
        if (data) {
          // 自动登录
          handler.postDelayed(mNavigateToMainTask, 500);
        } else {
          // 显示登录和注册按钮
          view.showButtons();
        }
      }
    });
  }

  private Runnable mNavigateToMainTask = new Runnable() {
    @Override
    public void run() {
      view.navigateToMain();
    }
  };

  @Override
  public void takeView(SplashContract.View view) {
    this.view = view;
  }

  @Override
  public void dropView() {
    this.view = null;
    autoLogin.unsubscribe();
    handler.removeCallbacks(mNavigateToMainTask);
  }
}

File: app-presentation/src/main/java/com/zxjdev/smile/presentation/infrastucture/splash/SplashPresenter.kt
package com.zxjdev.smile.presentation.infrastucture.splash

import android.os.Handler
import com.zxjdev.smile.domain.authorization.usecase.AutoLogin
import com.zxjdev.smile.presentation.common.util.ui.ErrorMessagePrinter
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

class SplashPresenter @Inject internal constructor() : SplashContract.Presenter {

    @Inject lateinit var autoLogin: AutoLogin
    private lateinit var view: SplashContract.View
    @Inject lateinit var errorMessagePrinter: ErrorMessagePrinter
    private val handler: Handler = Handler()
    private val compositeDisposable = CompositeDisposable()

    private val mNavigateToMainTask = Runnable { view.navigateToMain() }

    override fun handleAutoLogin() {
        compositeDisposable.add(autoLogin.execute().subscribeWith(object : DisposableObserver<Boolean>() {
            override fun onComplete() {
            }

            override fun onNext(t: Boolean) {
                if (t) {
                    // 自动登录
                    handler.postDelayed(mNavigateToMainTask, 500)
                } else {
                    // 显示登录和注册按钮
                    view.showButtons()
                }
            }

            override fun onError(e: Throwable) {
                errorMessagePrinter.print(e.message)
            }
        }))
    }

    override fun takeView(view: SplashContract.View) {
        this.view = view
    }

    override fun dropView() {
        compositeDisposable.dispose()
        handler.removeCallbacks(mNavigateToMainTask)
    }
}
